### PR TITLE
Add odometry to markhor

### PR DESCRIPTION
**Pull request by benmalenfant**

### CODE REVIEW

File: markhor_tracks/include/markhor_hw_interface.hpp [hunk 1]
```diff
@@ -69,14 +69,16 @@ class MarkhorHWInterface : public hardware_interface::RobotHW
 
   double tracks_fb_coeff = 1;
 
+  double track_encoder_to_rad_coeff = 39384;
```
Comment (saxtot) [inline markhor_tracks/include/markhor_hw_interface.hpp:72]: Could be a constant, like the one on line 36

File: markhor_tracks/src/markhor_hw_interface.cpp [hunk 2]
```diff
@@ -248,6 +252,19 @@ void MarkhorHWInterface::write()
 
 void MarkhorHWInterface::read(const ros::Duration& period)
 {
+
+  //{ "flipper_fl_motor_j", "flipper_fr_motor_j", "flipper_rl_motor_j", "flipper_rr_motor_j" };
+
+  joint_position_[0] = -2*3.14159265358*front_left_drive->GetSensorCollection().GetPulseWidthPosition() / (double)track_encoder_to_rad_coeff;
```
Comment (saxtot) [inline markhor_tracks/src/markhor_hw_interface.cpp:258]: Pi probably exists somewhere as a constant or could be created.  Maybe 2*pi could be a constant with a representative name, it's not instantly clear what it is.
Be aware that every hard coded value ever will get this type of comment eventually.

File: markhor_tracks/src/markhor_hw_interface.cpp [hunk 3]
```diff
@@ -248,6 +252,19 @@ void MarkhorHWInterface::write()
 
 void MarkhorHWInterface::read(const ros::Duration& period)
 {
+
+  //{ "flipper_fl_motor_j", "flipper_fr_motor_j", "flipper_rl_motor_j", "flipper_rr_motor_j" };
```
Comment (saxtot) [inline markhor_tracks/src/markhor_hw_interface.cpp:256]: Necessary?

File: markhor_tracks/src/markhor_hw_interface.cpp [hunk 4]
```diff
@@ -248,6 +252,19 @@ void MarkhorHWInterface::write()
...
+  joint_position_[3] = 2*3.14159265358*rear_right_drive->GetSensorCollection().GetPulseWidthPosition() / (double)track_encoder_to_rad_coeff;
+
+  joint_velocity_[0] = -2*3.14159265358*front_left_drive->GetSensorCollection().GetPulseWidthVelocity() / (double)track_encoder_to_rad_coeff;
+  joint_velocity_[1] = 2*3.14159265358*front_right_drive->GetSensorCollection().GetPulseWidthVelocity() / (double)track_encoder_to_rad_coeff;
+  joint_velocity_[2] = -2*3.14159265358*rear_left_drive->GetSensorCollection().GetPulseWidthVelocity() / (double)track_encoder_to_rad_coeff;
+  joint_velocity_[3] = 2*3.14159265358*rear_right_drive->GetSensorCollection().GetPulseWidthVelocity() / (double)track_encoder_to_rad_coeff;
+
   // Read from the motor API, going to read from the TalonSRX objects
   // ROS_INFO("Vel: %d, %d",
   // rear_right_drive->GetSensorCollection().GetQuadratureVelocity(),rear_left_drive->GetSensorCollection().GetQuadratureVelocity());
```
Comment (saxtot) [inline markhor_tracks/src/markhor_hw_interface.cpp:270]: I know its not part of your changes, but would it possible to remove this if applicable?

File: markhor_tracks/include/markhor_hw_interface.hpp (hunk 1 above)
Comment (benmalenfant) [reply markhor_tracks/include/markhor_hw_interface.hpp:72]: Can't be a constant, this is the default value, this variable is overwritten by the parameter set in the launchfile

File: markhor_tracks/src/markhor_hw_interface.cpp (hunk 2 above)
Comment (benmalenfant) [reply markhor_tracks/src/markhor_hw_interface.cpp:258]: Good catch, I forgot it there, I'll use the pi defined in math.h

File: markhor_tracks/src/markhor_hw_interface.cpp (hunk 3 above)
Comment (benmalenfant) [reply markhor_tracks/src/markhor_hw_interface.cpp:256]: nope, was just a reminder while coding for the order of the flippers, I'll remove it

File: markhor_tracks/src/markhor_hw_interface.cpp (hunk 4 above)
Comment (benmalenfant) [reply markhor_tracks/src/markhor_hw_interface.cpp:270]: yeah, I'll remove it

**Review by saxtot — APPROVED** [review]
Great, works for me. Keep it up!

**Comment by benmalenfant** [comment]
Missing tracks.launch changes...

**Comment by benmalenfant** [comment]
Track system should be good to go with odometry! :D I re-tuned the pid right and it makes quite a difference, also had to tune down the max speeds as previous were just arbitrary and way larger since tracks were improperly calibrated.

@saxtot @GLDuval can we get this moving? This should allow further developpement of the autonomy @SimonR99

**Comment by saxtot** [comment]
So, what's "tune_tracks"?

File: markhor_bringup/config/xbox.config.yaml [hunk 5]
```diff
@@ -1,9 +1,9 @@
 axis_linear: 1
-scale_linear: 0.5
-scale_linear_turbo: 1.5
+scale_linear: 0.1
+scale_linear_turbo: 0.1
```
Comment (GLDuval) [inline markhor_bringup/config/xbox.config.yaml:3]: The scale_linear and scale_linear_turbo values are the same. Is this normal?

**Review by GLDuval — COMMENTED** [review]
Nice work!

**Comment by benmalenfant** [comment]
> So, what's "tune_tracks"?

A launch file to allow easy and precise tuning, basically just replaces the diff drive controller by passthrough velocity-velocity controllers, which you can dirrectly control via the command topic

**Review by saxtot — APPROVED** [review]
Still works for me!
Not the biggest fan of the "tune_tracks" name. "tracks_tuner" would be as simple and clearer in its intention, but it works.

**Review by saxtot — APPROVED** [review]
I've reviewed the last changes, but please start a new branch instead of pushing on this PR again.

**Review by GLDuval — APPROVED** [review]

**Review by GLDuval — APPROVED** [review]

**Review by SimonR99 — APPROVED** [review]

**Pull request merged**